query/src/test/java/com/connexta/replication/query/cmp/QueryServiceComponentTest.java

```diff
@@ -0,0 +1,421 @@
+    siteManager.save(site);
+    final Filter filter = new FilterImpl(generateFilterPojo("1", 11));
+    filterManager.save(filter);
+    List<Metadata> metadataList =
+        List.of(
+            generateMetadataWithResource(111),
+            generateMetadataWithResource(112),
+            generateMetadataWithResource(113, new Date(1)));
+    when(adapter.query(Mockito.argThat(isQueryRequestWithFilter(filter.getFilter()))))
+        .thenReturn(
+            new QueryResponseImpl(
+                metadataList,
+                e -> {
+                  throw new AssertionError(e);
+                }));
+
+    indexManager = Mockito.spy(indexManager);
+
+    startQueryManager(THIRTY_MINUTES);
+
+    final MemorySiteQueue queue = (MemorySiteQueue) broker.getQueue(site.getId());
+
+    assertThat(
+        "timed out waiting for expected published task",
+        queue.waitForPendingSizeToReach(3, 5L, TimeUnit.SECONDS),
+        Matchers.equalTo(true));
+
+    for (Metadata metadata : metadataList) {
+      assertTask(queue.take(), metadata, (byte) 0, OperationType.HARVEST);
+    }
+    verify(indexManager)
+        .save(
+            Mockito.argThat(
+                isIndexWithModified(metadataList.get(2).getMetadataModified().toInstant())));
+  }
+
+  @Test
+  public void testQueryThrough2Polls() throws Exception {
+    final Site site = new SiteImpl(generateSitePojo(1));
+    siteManager.save(site);
+    final Filter filter = new FilterImpl(generateFilterPojo("1", 11));
+    filterManager.save(filter);
+    List<Metadata> metadataList =
+        List.of(
+            generateMetadataWithResource(111),
+            generateMetadataWithResource(112),
+            generateMetadataWithResource(113, new Date(1)));
+    Metadata secondPollMetadata = generateMetadataWithResource(114, new Date(2));
+    when(adapter.query(Mockito.argThat(isQueryRequestWithFilter(filter.getFilter()))))
+        .thenReturn(
+            new QueryResponseImpl(
+                metadataList,
+                e -> {
+                  throw new AssertionError(e);
+                }),
+            new QueryResponseImpl(
+                List.of(secondPollMetadata),
+                e -> {
+                  throw new AssertionError(e);
+                }));
+
+    indexManager = Mockito.spy(indexManager);
+
+    startQueryManager(THIRTY_MINUTES);
+
+    final MemorySiteQueue queue = (MemorySiteQueue) broker.getQueue(site.getId());
+
+    assertThat(
+        "timed out waiting for expected published task",
+        queue.waitForPendingSizeToReach(3, 5L, TimeUnit.SECONDS),
+        Matchers.equalTo(true));
+    for (Metadata metadata : metadataList) {
+      assertTask(queue.take(), metadata, (byte) 0, OperationType.HARVEST);
+    }
+    verify(indexManager)
+        .save(
+            Mockito.argThat(
+                isIndexWithModified(metadataList.get(2).getMetadataModified().toInstant())));
+
+    for (final Task task : tasks) {
+      task.complete();
```

Comment:
❗️ There is a danger here of having a test that sometime will work and sometime might fail. This is because of the instability created by the first `task.complete()`. When the first one is called, the queue will automatically wakeup the thread waiting to push the 4th task in it. If that task has higher priority, it might end up being pulled as the second task instead of being the 4th. We need to ensure that this doesn't happen in a consistent way. We can talk about it later.